File: sdk-go/converge/serve.go
package converge

import (
	"context"
	"log/slog"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

// Serve is the reusable BODY of a converge worker binary: it configures logging, reads
// config from the environment (via New — BROKER_ADDR / TLS_* / WORKER_*, cloud-SDK
// style), registers the given providers (scoped by WORKER_KINDS), optionally starts the
// DB-free k8s probe listener (HEALTH_ADDR), then BLOCKS on the worker's Run until ctx is
// cancelled — draining in-flight work — and returns its error. It is the convenience
// shell (kinds filter + probes + logging) over New(...).Add(...).Run(ctx).
//
// The caller owns ctx and the exit decision: build the context (a SIGINT/SIGTERM
// signal.NotifyContext for a standalone binary, or a parent errgroup/request ctx when a
// worker is embedded in a bigger app) and act on the returned error. A standalone worker
// main is 4 lines:
//
//	func main() {
//		ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
//		defer stop()
//		if err := converge.Serve(ctx, []converge.Provider{myprov.Provider{}}); err != nil {
//			log.Fatal(err)
//		}
//	}
//
// A program that wants finer control skips Serve and drives the Worker directly
// (New(...).Add(...).Run(ctx)) — see examples/demos/datadriven/cmd/worker.
//
// providers each satisfy Provider (the ONE SDK provider contract). Returns a non-nil
// error if WORKER_KINDS matched none of the providers (nothing to run), or whatever
// Worker.Run returns.
func Serve(ctx context.Context, providers []Provider) error {
	setupLogger()
	// The worker reads BROKER_ADDR / TLS_* / WORKER_* from the environment itself
	// (cloud-SDK style) and owns dial + reconnect + config + readiness + drain.
	// Construction errors only on a malformed SET env var — surface it rather than run
	// mis-configured.
	w, err := newWorker()
	if err != nil {
		return err
	}

	// Register the providers this binary serves, scoped by WORKER_KINDS. Each provider
	// self-manages its bring-up and reports Ready — the SDK sends it work only for the
	// (kind, version) pair it advertises via Kind().
	want := kindFilter(os.Getenv("WORKER_KINDS"))
	added := 0
	for _, p := range providers {
		if want != nil && !providerWanted(p, want) {
			continue // no wanted kind → don't register
		}
		w.register(p)
		added++
	}
	if added == 0 {
		return errNoKinds(os.Getenv("WORKER_KINDS"))
	}

	// Optional DB-free k8s probe listener (a worker holds no DB, so liveness is a bare
	// 200 while the process is up; /readyz flips NotReady once shutdown begins so k8s
	// pulls the pod from Service endpoints during the drain).
	if addr := os.Getenv("HEALTH_ADDR"); addr != "" {
		serveProbes(ctx, addr)
	}

	slog.Info("worker starting", "providers", added)
	return w.run(ctx) // blocks until ctx is cancelled, then drains + returns
}

// serveProbes starts the DB-free k8s liveness/readiness listener on addr, bound to ctx.
func serveProbes(ctx context.Context, addr string) {
	mux := http.NewServeMux()
	live := func(rw http.ResponseWriter, _ *http.Request) { rw.WriteHeader(http.StatusOK) }
	mux.HandleFunc("/healthz", live)
	mux.HandleFunc("/livez", live)
	mux.HandleFunc("/readyz", func(rw http.ResponseWriter, _ *http.Request) {
		if ctx.Err() != nil { // shutdown began → NotReady
			rw.WriteHeader(http.StatusServiceUnavailable)
			return
		}
		rw.WriteHeader(http.StatusOK)
	})
	hs := &http.Server{Addr: addr, Handler: mux, ReadHeaderTimeout: 10 * time.Second}
	go func() {
		if err := hs.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			slog.Error("worker health listener", "err", err)
		}
	}()
}

// kindFilter parses the comma-separated WORKER_KINDS env into a set, or nil (all kinds).
func kindFilter(csv string) map[Kind]bool {
	csv = strings.TrimSpace(csv)
	if csv == "" {
		return nil
	}
	m := map[Kind]bool{}
	for k := range strings.SplitSeq(csv, ",") {
		if k = strings.TrimSpace(k); k != "" {
			m[Kind(k)] = true
		}
	}
	return m
}

// providerWanted reports whether a provider's kind is in want (reads the pure Kind(), no
// dial), so a provider whose kind isn't wanted is skipped without registering.
func providerWanted(p Provider, want map[Kind]bool) bool {
	return want[p.Kind().Kind]
}

// errNoKinds is the "nothing to run" error, kept as a named builder so the message stays
// in one place.
func errNoKinds(workerKinds string) error {
	return &noKindsError{workerKinds: workerKinds}
}

type noKindsError struct{ workerKinds string }

func (e *noKindsError) Error() string {
	return "converge: no providers to run (WORKER_KINDS=" + strconv.Quote(e.workerKinds) + " matched nothing)"
}

// setupLogger configures the default slog logger from LOG_LEVEL (a level name or a
// numeric slog level) + LOG_FORMAT (text|json). Defaults: info / text.
func setupLogger() {
	lvl := slog.LevelInfo
	if s := os.Getenv("LOG_LEVEL"); s != "" {
		if l, err := strconv.Atoi(s); err == nil {
			lvl = slog.Level(l)
		} else {
			switch strings.ToLower(s) {
			case "debug":
				lvl = slog.LevelDebug
			case "warn":
				lvl = slog.LevelWarn
			case "error":
				lvl = slog.LevelError
			}
		}
	}
	opts := &slog.HandlerOptions{Level: lvl}
	var h slog.Handler = slog.NewTextHandler(os.Stderr, opts)
	if strings.EqualFold(os.Getenv("LOG_FORMAT"), "json") {
		h = slog.NewJSONHandler(os.Stderr, opts)
	}
	slog.SetDefault(slog.New(h))
}

File: sdk-ts/src/serve.ts
// serve.ts is the HIGH-LEVEL worker entrypoint. It reads config from the ENVIRONMENT
// (BROKER_ADDR / TLS_* / WORKER_* — the "zero-config" path), registers the given providers
// (scoped by WORKER_KINDS), optionally starts the k8s probe listener (HEALTH_ADDR), dials
// the broker, and runs the WorkStream loop until the AbortSignal fires — draining in-flight
// work. It is the convenience shell over the same run loop the low-level runWorker drives.
//
// The caller owns the signal + the exit decision. A worker main is: build an AbortController
// wired to SIGINT/SIGTERM, await serve(providers, controller.signal). serve reads nothing but
// the environment — there is no options argument (use runWorker for programmatic control).
import { createServer, type Server } from "node:http";
import { loadConfig, newLogger, type Config, type Logger } from "./config.js";
import { runLoop } from "./runner.js";
import { dial, watchTLS } from "./transport.js";
import type { Provider } from "./types.js";

// serve runs the worker's providers until signal aborts, reading ALL config from the
// environment. It BLOCKS (returns a Promise that resolves on clean shutdown). Throws on a
// malformed env var (loud, never a silent fallback), if WORKER_KINDS matched none of the
// providers (nothing to run), or on a fatal run error.
export async function serve(providers: Provider[], signal: AbortSignal): Promise<void> {
  const log = newLogger();
  const cfg = loadConfig();

  // Register the providers this binary serves, scoped by WORKER_KINDS. A provider whose kind
  // isn't in the (non-null) allowlist is dropped without connecting.
  const active = cfg.workerKinds === null ? providers : providers.filter((p) => cfg.workerKinds!.has(p.kind().kind));
  if (active.length === 0) {
    throw new Error(`converge: no providers to run (WORKER_KINDS matched nothing)`);
  }
  if (cfg.brokerAddr === "") {
    throw new Error("converge: BROKER_ADDR is required");
  }

  // Optional k8s liveness/readiness probes (a worker holds no DB, so liveness is a bare 200
  // while up; /readyz flips NotReady once shutdown begins so k8s pulls the pod during drain).
  const probes = cfg.healthAddr !== "" ? serveProbes(cfg.healthAddr, log, signal) : undefined;

  // Dial the broker; rebuild the client on a cert rotation (watchTLS) so mTLS material
  // hot-reloads with no restart. On the cleartext path watchTLS is a no-op.
  let client = dial(cfg);
  watchTLS(cfg, log, () => {
    client = dial(cfg);
  }, signal);

  log.info("worker starting", { providers: active.length, brokerAddr: cfg.brokerAddr });
  try {
    await runLoop(client, active, cfg, log, signal);
  } finally {
    probes?.close();
  }
}

// serveProbes starts the DB-free k8s liveness/readiness listener on addr, bound to signal.
// /livez + /healthz are a bare 200 while the process is up; /readyz flips 503 once shutdown
// begins (signal aborted) so k8s pulls the pod from Service endpoints during the drain.
function serveProbes(addr: string, log: Logger, signal: AbortSignal): Server {
  const { host, port } = splitAddr(addr);
  const server = createServer((req, res) => {
    if (req.url === "/readyz") {
      res.writeHead(signal.aborted ? 503 : 200).end();
      return;
    }
    // /livez, /healthz, anything else: alive while the process is up.
    res.writeHead(200).end();
  });
  server.listen(port, host, () => log.info("worker health listener started", { addr }));
  server.on("error", (err) => log.error("worker health listener", { err: String(err) }));
  signal.addEventListener("abort", () => server.close(), { once: true });
  return server;
}

// splitAddr parses a ":8080" or "host:8080" HEALTH_ADDR into { host, port }. A bare ":port"
// binds all interfaces (host undefined → Node's default).
function splitAddr(addr: string): { host: string | undefined; port: number } {
  const i = addr.lastIndexOf(":");
  const host = i > 0 ? addr.slice(0, i) : undefined;
  const port = Number(addr.slice(i + 1));
  return { host, port };
}

// Config/Logger are re-exported so an embedder can type against them if they build their
// own transport for runWorker; serve itself needs no options.
export type { Config, Logger };
